src/pushsource/_impl/backend/staged/staged_source.py

```diff
@@ -136,6 +137,25 @@ def _push_items_for_leafdir(self, leafdir, metadata):
     def _push_items_for_topdir(self, topdir):
         LOG.info("Checking files in: %s", topdir)
 
+        timeout = int(os.getenv("PUSHSOURCE_SRC_POLL_TIMEOUT") or "0")
+        poll_rate = int(os.getenv("PUSHSOURCE_SRC_POLL_RATE") or "30")
+
+        # If staged path doesn't exists, we can assume that once it does, it will
+        # contain fresh content. It's sufficient to only wait until it appears.
+        if not os.path.exists(topdir):
+            wait_exist(topdir, timeout, poll_rate)
```

Comment:
I don't know exactly how the replication mechanism works, but this logic seems a bit suspicious because it will only work under the assumption that, if a parent directory appears, all subdirectories have been fully replicated.  It does not seem a safe assumption without diving into exactly how it works.

I mean for example, looking at https://release-engineering.github.io/pushsource/sources/staged.html#staging-structure , can it be the case that:

- initially "root/" does not exist
- we poll for it to exist, and find that it eventually appears
- then we proceed through the `for entry in scandir(topdir)` link below
- but only half of the 'destination' dirs have replicated, so scandir misses half of the staged content

Is it possible to get a meaningful `mtime` of the topdir?  If so, the logic could be something like:

Check the mtime.  Sleep until the current time is at least (mtime + timeout).  Imagining a timeout of 15 minutes, it means if a directory was initially replicated just as we get up to this code, then we'll wait 15 minutes to hopefully ensure that all of the subdirectories are also fully replicated.

Reply:
I don't like the idea that every push-staged would have a 15 minutes delay at the beginning even if, for example, the user is pushing a staging area to production which was already pushed to stage hours/days earlier (which I think is a common case).  We'd probably get bug reports/complaints about that straight away.

How about this logic then:

- check the mtime of the topdir
- if and only if mtime falls in the range of (current_time - X minutes) .. (current_time), then apply the delay
- the value of X is configurable by env var

If we start by setting X to 24 hours, then the overall behavior will be: if we process a dir created within the last day, we need to delay because we assume it might still be changing.  If we process a dir created earlier than that, we assume it's stable.  And since it's configurable, if we chose a bad value of X we can tune it without code changes.  Putting a very large number will get us back to "delay every time" if it turns out this is really necessary.

Also, if we're doing anything along those lines it'd probably be best to have an INFO-level log message with the mtime of the directory, so that we can later reason about how much it should have delayed.

Reply:
Changed the logic to your proposal.